### PR TITLE
Pinpoint sms integration

### PR DESCRIPTION
**Pull request by ShivaniMehendarge**
*Issue #, if available:*

*Description of changes:*
Made changes to the pinpoint auto workshop to add:
1. Creating a segment for all SMS users.
2. An event based campaign that is triggered when the SMS user replies with Y to the confirmation message
3. An API handler in user service to create an endpoint for the sms user and send confirmation message to the user. Also added permissions required for this service to use pinpoint.
4. an SNS topic to which the registered long code will send message to
5. A lambda that is triggered by the SNS topic. It updates the SMS endpoint for the user and triggers the event for the Campaign.
6. an SSM parameter for long code - to be set manually by the user.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/web-ui/src/public/Footer.vue [hunk 1]
```diff
@@ -1,6 +1,19 @@
 <template>
   <div class="footer">
     <div class="container">
+      <div class="row" v-if="user">
+        <form class="mt-2 mb-2 mx-auto text-center">
+          <div class="form-group mb-2 mr-sm-2">
+            <label for="sms-alerts">Join Text Alerts and get 20% off</label>
```
Comment (alexchirayath) [inline src/web-ui/src/public/Footer.vue:7]: Just curious: How do you check if Pinpoint is enbaled before showing the text alerts?

File: src/web-ui/src/public/Footer.vue [hunk 2]
```diff
@@ -12,9 +25,38 @@
 </template>
 
 <script>
+import { RepositoryFactory } from '@/repositories/RepositoryFactory'
+//import { AnalyticsHandler } from '@/analytics/AnalyticsHandler'
```
Comment (alexchirayath) [inline src/web-ui/src/public/Footer.vue:29]: Could you remove comments?

File: stage.sh [hunk 3]
```diff
@@ -12,7 +12,7 @@ BUCKET=$1
 S3PATH=$2
 
 # remove this line if you want to keep the objects private in your S3 bucket
-export S3PUBLIC=" --acl public-read"
+#export S3PUBLIC=" --acl public-read"
```
Comment (alexchirayath) [inline stage.sh:15]: I think you missed to uncomment this piece

File: src/web-ui/.env [hunk 4]
```diff
@@ -1,27 +1,27 @@
...
+VUE_APP_SEARCH_SERVICE_PORT=80
 
 # Configure Authentication
 VUE_APP_AWS_REGION=us-west-2
-VUE_APP_AWS_IDENTITY_POOL_ID=
-VUE_APP_AWS_USER_POOL_ID=
-VUE_APP_AWS_USER_POOL_CLIENT_ID=
+VUE_APP_AWS_IDENTITY_POOL_ID=us-west-2:47830704-39b5-4258-94f8-cb0efeed73dd
+VUE_APP_AWS_USER_POOL_ID=us-west-2_q4epyIDJp
+VUE_APP_AWS_USER_POOL_CLIENT_ID=1u5em88069h1fphhtti77e2qdb
```
Comment (james-jory) [inline src/web-ui/.env:24]: Do not check these changes in. Remove your updates to this file but keep the ones from the merge.

File: src/web-ui/src/main.js [hunk 5]
```diff
@@ -72,7 +72,7 @@ require('dotenv').config()
 
 Vue.config.productionTip = false
 
-Logger.LOG_LEVEL = 'DEBUG'
+//Logger.LOG_LEVEL = 'DEBUG'
```
Comment (james-jory) [inline src/web-ui/src/main.js:75]: Why this change?

File: src/web-ui/src/public/Footer.vue [hunk 6]
```diff
@@ -1,6 +1,19 @@
...
     <div class="container">
+      <div class="row" v-if="user">
+        <form class="mt-2 mb-2 mx-auto text-center">
+          <div class="form-group mb-2 mr-sm-2">
+            <label for="sms-alerts">Join Text Alerts and get 20% off</label>
+            <div class="input-group">
+              <input type="text" class="form-control" id="sms-alerts" placeholder="Enter cell phone number" aria-describedby="basic-addon2">
+              <div class="input-group-append">
+                <button type="button" class="btn btn-outline-secondary mb-2" v-on:click="verifyAndUpdateUserPhoneNumber">Submit</button>
+              </div>
```
Comment (james-jory) [inline src/web-ui/src/public/Footer.vue:12]: I assume the messaging for "messaging and data rates may apply..." will be added at some point?

File: src/web-ui/src/public/Footer.vue (hunk 2 above)
Comment (james-jory) [inline src/web-ui/src/public/Footer.vue:29]: Remove commented code

File: src/web-ui/src/public/Footer.vue [hunk 7]
```diff
@@ -12,9 +25,38 @@
...
+      if (userID && userID.length > 0) {
+        const { data } = await UsersRepository.getUserByID(userID)
+        this.user = data
+      }
+      return this.user
+    },
+    async verifyAndUpdateUserPhoneNumber() {
+      var userPhoneNumber = document.getElementById('sms-alerts').value
+      const { data } = await UsersRepository.verifyAndUpdateUserPhoneNumber(AmplifyStore.state.user.id, userPhoneNumber)
+      console.log(data)
```
Comment (james-jory) [inline src/web-ui/src/public/Footer.vue:57]: Should we have some better error handling here? Like displaying a message when a bad phone number is entered and the API returns a response telling us so.

File: src/web-ui/src/repositories/usersRepository.js [hunk 8]
```diff
@@ -45,5 +45,14 @@ export default {
         if (!user)
             throw "user required"
         return connection.put(`${resource}/id/${user.id}`, user)
+    },
+    verifyAndUpdateUserPhoneNumber(userID, phonenumber) {
+        if (!userID || userID.length == 0)
+            throw "userID required"
+        let payload = {
+            userid: userID,
+            phonenumber: phonenumber
```
Comment (james-jory) [inline src/web-ui/src/repositories/usersRepository.js:54]: Sync these changes with requested naming changes earlier in PR (i.e. snake case).

File: src/web-ui/src/repositories/usersRepository.js [hunk 9]
```diff
@@ -45,5 +45,14 @@ export default {
...
         return connection.put(`${resource}/id/${user.id}`, user)
+    },
+    verifyAndUpdateUserPhoneNumber(userID, phonenumber) {
+        if (!userID || userID.length == 0)
+            throw "userID required"
+        let payload = {
+            userid: userID,
+            phonenumber: phonenumber        
+        }
+        return connection.post(`${resource}/id/${userID}/verifyphone`, payload)
```
Comment (james-jory) [inline src/web-ui/src/repositories/usersRepository.js:56]: Should be `put()` per earlier comment.

File: src/web-ui/src/router/index.js [hunk 10]
```diff
@@ -110,6 +110,9 @@ AmplifyEventBus.$on('authState', async (state) => {
       storeUser.sign_up_date = now.toISOString()
       newSignUp = true
     }
+    if (!storeUser.phone_number) {
+      storeUser.phone_number = null
+    }
```
Comment (james-jory) [inline src/web-ui/src/router/index.js:115]: This doesn't make sense. Why not always set the phone number to null? Why set it to null in the first place? Deserves a comment at least, if even necessary.

File: stage.sh (hunk 3 above)
Comment (james-jory) [inline stage.sh:15]: Did you intend to change the default behavior here?

File: workshop/4-Messaging/4.1-Pinpoint.ipynb [hunk 11]
```diff
@@ -691,7 +691,7 @@
     "\n",
     "![Create Retail Demo Store account](images/retaildemostore-create-account.png)\n",
     "\n",
-    "3. A few seconds after creating your account you should receive an SMS with a six digit confirmation code. Enter this code on the confirmation page.\n",
+    "3. A few seconds after creating your account you should receive an email with a six digit confirmation code. Enter this code on the confirmation page.\n",
```
Comment (james-jory) [inline workshop/4-Messaging/4.1-Pinpoint.ipynb:694]: This notebook needs to be updated to include all the functionality in the pre-create lambda function. In other words, creating the segment, campaign, and message template for the SMS functionality.

**Review by james-jory — CHANGES_REQUESTED** [review]

**Comment by james-jory** [comment]
I noticed you had HTML meta tags for the product detail view but removed them. Were they not needed to preview product details in mobile phones?

**Comment by ShivaniMehendarge** [comment]
> I noticed you had HTML meta tags for the product detail view but removed them. Were they not needed to preview product details in mobile phones?

SMS does not render HTML tags. It just displayed them as  plain text. To display the image preview, we need to add meta tags to the <head> of ProductDetail page and also add server side rendering to RDS. Just adding the meta tags did not work.

File: workshop/4-Messaging/4.1-Pinpoint.ipynb (hunk 11 above)
Comment (ShivaniMehendarge) [reply workshop/4-Messaging/4.1-Pinpoint.ipynb:694]: Yes, I will update the workshop.

File: stage.sh (hunk 3 above)
Comment (ShivaniMehendarge) [reply stage.sh:15]: No, I forgot to uncomment it. Had commented it because I was using my own s3 bucket for deployment.

File: src/web-ui/src/router/index.js (hunk 10 above)
Comment (ShivaniMehendarge) [reply src/web-ui/src/router/index.js:115]: Good catch James. I need to remove this piece. This was added as a part of the trial and error when the event from UI was not triggering the campaign.

File: stage.sh (hunk 3 above)
Comment (ShivaniMehendarge) [reply stage.sh:15]: Yep. Good point.

File: aws/cloudformation-templates/base/authentication.yaml [hunk 12]
```diff
@@ -46,19 +46,11 @@ Resources:
       MfaConfiguration: "OFF"
       AutoVerifiedAttributes:
         - email
-        - phone_number
-      SmsConfiguration:
-        ExternalId: !Sub ${AuthName}-external
-        SnsCallerArn: !GetAtt SNSRole.Arn
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/base/authentication.yaml:52]: Let me check if it is used anywhere else. If not, I will remove it. Thanks for pointing it out.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 13]
```diff
@@ -503,3 +503,95 @@ Resources:
         Fn::GetAtt: 
           - "PinpointPreCreateRule"
           - "Arn"
+
+  ####################### Pre-Create Pinpoint SMS Alerts Lambda #######################
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/deployment-support.yaml:507]: I saw that the Pinpoint Auto workshop is in this template itself and hence added it here. But I can surely shift it to another template.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 14]
```diff
@@ -503,3 +503,95 @@ Resources:
...
+  PinpointSMSAlertsLambda:
+    Condition: DeployPreCreatePinpointWorkshop
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/deployment-support.yaml:509]: Good point! I got it confused with being similar to PinpointAutoWorkshop lambda. I will remove the condition, since the lambda needs to be present for the customer to run it using the workshop.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 15]
```diff
@@ -503,3 +503,95 @@ Resources:
...
+                  - !Sub 'arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/lambda/RetailDemoStorePinpointSMSAlerts'
+              - Effect: Allow
+                Action:
+                  - ssm:GetParameter
+                Resource: !Sub 'arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/retaildemostore-*'
+              - 
+                Effect: "Allow"
+                Action: 
+                  - mobiletargeting:*
+                Resource: "*"
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/deployment-support.yaml:566]: I'll try scoping them down again, but last I tried it gave me permissions error.

File: aws/cloudformation-templates/deployment-support.yaml [hunk 16]
```diff
@@ -503,3 +503,95 @@ Resources:
...
+              - Effect: Allow
+                Action:
+                  - events:ListTargetsByRule
+                  - events:RemoveTargets
+                  - events:DeleteRule
+                Resource:
+                  - !Sub 'arn:aws:events:${AWS::Region}:${AWS::AccountId}:rule/RetailDemoStore-PinpointPreCreateRule'
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/deployment-support.yaml:573]: Good catch! I will remove these permissions. They are not needed.

File: src/aws-lambda/pinpoint-sms-alerts/pinpoint-sms-alerts.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    pinpoint_app_id = os.environ['pinpoint_app_id']
+
+    timestamp =  event['Records'][0]['Sns']['Timestamp']
+    message = eval(event['Records'][0]['Sns']['Message'])
+    originationNumber = message['originationNumber']
+    response = message['messageBody'].lower()
+    endpointId = originationNumber[1:]
+
+    # opt in the customer to receive text alerts if they responded with 'y' to the confirmation SMS
+    if 'y' in response:
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply src/aws-lambda/pinpoint-sms-alerts/pinpoint-sms-alerts.py:62]: Agreed. I will update it.

File: src/users/src/users-service/handlers.go [hunk 18]
```diff
@@ -179,6 +182,144 @@ func UserCreate(w http.ResponseWriter, r *http.Request) {
...
+func UserVerifyAndUpdatePhone(w http.ResponseWriter, r *http.Request){
+	enableCors(&w)
+	if (*r).Method == "OPTIONS" {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+	type Useridandnumber struct {
+		Userid		string 	   `json:"userid" yaml:"userid"`
+		PhoneNumber string 	   `json:"phonenumber" yaml:"phonenumber"`
+	}
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply src/users/src/users-service/handlers.go:238]: I tried using this naming convention and for some reason Go did not unmarshal the request into the object. I changed it to this naming convention after coming across a stackoverflow answer. I can verify it once more.

File: src/users/src/users-service/handlers.go [hunk 19]
```diff
@@ -179,6 +182,144 @@ func UserCreate(w http.ResponseWriter, r *http.Request) {
...
+		}
+		phoneValidateInput := &pinpoint.PhoneNumberValidateInput{
+			NumberValidateRequest: numberValidaterequest,
+		}
+		res, err := pinpoint_client.PhoneNumberValidate(phoneValidateInput)
+		if err != nil {
+			fmt.Println("Got error calling PhoneNumberValidate:")
+			fmt.Println(err.Error())
+		} else {
+				fmt.Println(res)
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply src/users/src/users-service/handlers.go:274]: I think yes. Good catch. Will update

File: src/web-ui/.env (hunk 4 above)
Comment (ShivaniMehendarge) [reply src/web-ui/.env:24]: Oops. I missed to remove this file. Sorry about that.

File: src/web-ui/src/main.js (hunk 5 above)
Comment (ShivaniMehendarge) [reply src/web-ui/src/main.js:75]: We were just trying something for the image preview to show up. But it did not work and I forgot to uncomment this piece. This needs to be uncommented.

File: src/web-ui/src/public/Footer.vue (hunk 6 above)
Comment (ShivaniMehendarge) [reply src/web-ui/src/public/Footer.vue:12]: Yes, I need to update this section with Kaz's changes. I will update it with the rest of the changes today.

File: src/web-ui/src/public/Footer.vue (hunk 7 above)
Comment (ShivaniMehendarge) [reply src/web-ui/src/public/Footer.vue:57]: Good point. I can add it.

File: src/aws-lambda/pinpoint-auto-workshop/pinpoint-auto-workshop.py [hunk 20]
```diff
@@ -477,6 +597,31 @@ def lambda_handler(event, context):
     campaign_config = create_abandoned_cart_campaign(pinpoint_app_id, email_from, users_with_cart_segment_id, users_with_cart_segment_version)
     logger.debug(json.dumps(campaign_config, indent = 2, default = str))
 
+    # Create SMS template and enable SMS channel
+    response = ssm.get_parameter(Name='retaildemostore-pinpoint-sms-longcode')
+    pinpoint_sms_long_code = response['Parameter']['Value']
+
+    assert pinpoint_sms_long_code != 'NONE', 'Pinpoint SMS long code value not set'
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply src/aws-lambda/pinpoint-auto-workshop/pinpoint-auto-workshop.py:604]: As discussed offline, I will remove the assert statement

File: src/users/src/users-service/handlers.go [hunk 21]
```diff
@@ -270,44 +262,49 @@ func UserVerifyAndUpdatePhone(w http.ResponseWriter, r *http.Request){
...
-						EndpointId: &endpointId,
-						EndpointRequest: endpointRequest,
-					}
-					CreateEndpointAndSendConfirmation(updateEndpointInput, userdetails.PhoneNumber)
+			fmt.Println(res)
+			mobilePhoneCode := int(*res.NumberValidateResponse.PhoneTypeCode)
+			if (mobilePhoneCode != 0) {
+				var errMessage string = "The phone number provided is not a MOBILE phone number. The number is not capable of receiving SMS. Cannot create SMS endpoint for this number. Try entering a mobile phone number."
+				panic(errMessage)
+				http.Error(w, errMessage, 500)
```
Comment (james-jory) [inline src/users/src/users-service/handlers.go:273]: 500 doesn't seem like the right error code here. The caller provided a bad phone number so something like a 422 or 400 would be more appropriate.

File: src/web-ui/src/repositories/usersRepository.js [hunk 22]
```diff
@@ -50,9 +50,9 @@ export default {
         if (!userID || userID.length == 0)
             throw "userID required"
         let payload = {
-            userid: userID,
-            phonenumber: phonenumber        
+            user_id: userID,
+            phone_number: phonenumber
```
Comment (james-jory) [inline src/web-ui/src/repositories/usersRepository.js:54]: Naming standard for client side JS is camel case. So `userId` and `phoneNumber` is preferred.

File: workshop/4-Messaging/4.1-Pinpoint.ipynb [hunk 23]
```diff
@@ -817,6 +817,172 @@
...
+    "Here we will create SMS template that includes recommendations from the Amazon Personalize product recommendation campaign that we created in the Personalization workshop. If you haven't completed the personalization workshop, please do so now and come back to this workshop when complete. The body of the SMS message is passed in as text while creating the sms template as shown below."
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "source": [],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
```
Comment (james-jory) [inline workshop/4-Messaging/4.1-Pinpoint.ipynb:846]: Empty cell? Needed?

File: workshop/4-Messaging/4.1-Pinpoint.ipynb [hunk 24]
```diff
@@ -817,6 +817,172 @@
...
+    "\n",
+    "![Text alerts section] (images/pinpoint-text-alerts-ui.png) \n",
+    "\n",
+    "* Enter your phone number along with the \"+\" and country code (without any spaces). Eg. \"+1XXXXXXXXXX\"\n",
+    "* This will send a confirmation SMS message to the phone number entered above, asking the user to confirm if they want to receive a one time SMS from Retail Demo Store. \n",
+    "* Reply with a Y in order to opt in for receiving messages. If you reply with anything apart from Y, the SMS endpoint will be created, but it will not be opted in i.e. the segment and campaigns that we create in the next steps, will not send messages to the user unless they have opted in for the messages.\n",
+    "* Once you reply with Y, the endpoint is opted in to receive messages from Pinpoint.\n",
+    "* One user can have only 1 registered SMS endpoint even if they enter multiple phone numbers. The number you entered the last will be taken as the final one.\n",
+    "* If you wish to create multiple endpoints, sign in using different users and enter different phone numbers for every user."
```
Comment (james-jory) [inline workshop/4-Messaging/4.1-Pinpoint.ipynb:947]: This needs to be done in separate browsers or Incognito/Private windows, right? Should that be added to the instructions?

File: workshop/4-Messaging/4.1-Pinpoint.ipynb [hunk 25]
```diff
@@ -817,6 +817,172 @@
...
+   "source": [
+    "## Create Retail Semo Store User Accounts & Amazon Pinpoint SMS Endpoints\n",
+    "Next, refer to the previous section of this workshop to create a new user account if you haven't done so already. Once you have signed in as a Retail Demo Store user, the text alerts section is shown near the footer of the app as shown in the figure below. \n",
+    "\n",
+    "![Text alerts section] (images/pinpoint-text-alerts-ui.png) \n",
+    "\n",
+    "* Enter your phone number along with the \"+\" and country code (without any spaces). Eg. \"+1XXXXXXXXXX\"\n",
+    "* This will send a confirmation SMS message to the phone number entered above, asking the user to confirm if they want to receive a one time SMS from Retail Demo Store. \n",
+    "* Reply with a Y in order to opt in for receiving messages. If you reply with anything apart from Y, the SMS endpoint will be created, but it will not be opted in i.e. the segment and campaigns that we create in the next steps, will not send messages to the user unless they have opted in for the messages.\n",
+    "* Once you reply with Y, the endpoint is opted in to receive messages from Pinpoint.\n",
```
Comment (james-jory) [inline workshop/4-Messaging/4.1-Pinpoint.ipynb:945]: The Lambda makes this all work. Where is that described in the notebook?

**Review by james-jory — CHANGES_REQUESTED** [review]
Noticed you added some new screenshots to the Pinpoint workshop. Did you test deployment of this branch to make sure the SM notebook instance deploys cleanly? We're already super close to the max 20MB archive size so just want to make sure these new screenshots don't put us over the limit.

Other detailed comments in my review.

File: aws/cloudformation-templates/deployment-support.yaml (hunk 15 above)
[the remark this answers is not in the document]
Comment (alexchirayath) [reply aws/cloudformation-templates/deployment-support.yaml]: It seems that we can at least scope it down to the region and account id?
https://docs.aws.amazon.com/pinpoint/latest/developerguide/permissions-actions.html

File: workshop/4-Messaging/4.1-Pinpoint.ipynb (hunk 24 above)
Comment (alexchirayath) [reply workshop/4-Messaging/4.1-Pinpoint.ipynb:947]: +1

File: aws/cloudformation-templates/template.yaml [hunk 26]
```diff
@@ -174,6 +177,12 @@ Parameters:
       is set to 'Yes'.
     Default: "AWS Retail Demo Store"
   
+  PinpointSMSLongCode:
+    Type: String
+    Description: >
+      Phone number(a dedicated long code) for Amazon Pinpoint to send and receive messages at. Be sure to enable two way SMS for this long code using Pinpoint. This parameter is required when 'Auto-Configure Pinpoint'
+      is set to 'Yes'.
```
[the remark this answers is not in the document]
Comment (alexchirayath) [reply aws/cloudformation-templates/template.yaml:184]: @james-jory and @shivani  : Do you think it would make sense to also have an option to include/exclude SMS integration for Pinpoint?

**Comment by ShivaniMehendarge** [comment]
> Noticed you added some new screenshots to the Pinpoint workshop. Did you test deployment of this branch to make sure the SM notebook instance deploys cleanly? We're already super close to the max 20MB archive size so just want to make sure these new screenshots don't put us over the limit.
> 
> Other detailed comments in my review.

Yes it is 19.1MB at the moment. I just added the screenshots that I felt were necessary. I can remove some if not needed.

File: workshop/4-Messaging/4.1-Pinpoint.ipynb (hunk 25 above)
Comment (ShivaniMehendarge) [reply workshop/4-Messaging/4.1-Pinpoint.ipynb:945]: I'll add it.

File: workshop/4-Messaging/4.1-Pinpoint.ipynb (hunk 24 above)
Comment (ShivaniMehendarge) [reply workshop/4-Messaging/4.1-Pinpoint.ipynb:947]: Yes I will add it.

File: workshop/4-Messaging/4.1-Pinpoint.ipynb (hunk 23 above)
Comment (ShivaniMehendarge) [reply workshop/4-Messaging/4.1-Pinpoint.ipynb:846]: No, must have been added by mistake. Will remove.

File: src/web-ui/src/repositories/usersRepository.js (hunk 22 above)
Comment (ShivaniMehendarge) [reply src/web-ui/src/repositories/usersRepository.js:54]: Will update.

File: src/users/src/users-service/handlers.go (hunk 21 above)
Comment (ShivaniMehendarge) [reply src/users/src/users-service/handlers.go:273]: Okay, I will change it to 422

File: aws/cloudformation-templates/services/pinpoint-personalize.yaml [hunk 27]
```diff
@@ -93,6 +96,82 @@ Resources:
       FunctionName: !GetAtt CustomizeRecommendationsFunction.Arn
       Principal: pinpoint.amazonaws.com
       SourceArn: !Sub 'arn:aws:mobiletargeting:${AWS::Region}:${AWS::AccountId}:*'
+  
+  ####################### Pre-Create Pinpoint SMS Alerts Lambda #######################
+  PinpointSMSAlertsLambda:
+    Type: 'AWS::Lambda::Function'
+    Properties:
+      Description: 'Retail Demo Store deployment utility function that opts in customers to receive sms alerts from Amazon Pinpoint'
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/services/pinpoint-personalize.yaml:104]: Will update.

File: aws/cloudformation-templates/services/pinpoint-personalize.yaml [hunk 28]
```diff
@@ -93,6 +96,82 @@ Resources:
       FunctionName: !GetAtt CustomizeRecommendationsFunction.Arn
       Principal: pinpoint.amazonaws.com
       SourceArn: !Sub 'arn:aws:mobiletargeting:${AWS::Region}:${AWS::AccountId}:*'
+  
+  ####################### Pre-Create Pinpoint SMS Alerts Lambda #######################
```
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/services/pinpoint-personalize.yaml:100]: Right. I'll remove it.

File: aws/cloudformation-templates/template.yaml (hunk 26 above)
[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/template.yaml:184]: I think if we make it optional, it is up to the customers to opt in for SMS integration right? @james-jory I set the default value of the long code to NONE. We are anyways checking that the long code is not NONE in the auto workshop before creating the segment, campaign and template.

**Comment by james-jory** [comment]
> > I noticed you had HTML meta tags for the product detail view but removed them. Were they not needed to preview product details in mobile phones?
> 
> SMS does not render HTML tags. It just displayed them as plain text. To display the image preview, we need to add meta tags to the of ProductDetail page and also add server side rendering to RDS. Just adding the meta tags did not work.

Okay, that makes sense. Maybe this is something we can come back to later. It would be nice to get SMS previews working for product links. SSR would be a major change and I'm not it would be worth the effort. Perhaps a L@E function could be used to  inject the open graph meta tags when clients access the app by deep linking to product detail pages.

**Comment by james-jory** [comment]
> > Noticed you added some new screenshots to the Pinpoint workshop. Did you test deployment of this branch to make sure the SM notebook instance deploys cleanly? We're already super close to the max 20MB archive size so just want to make sure these new screenshots don't put us over the limit.
> > Other detailed comments in my review.
> 
> Yes it is 19.1MB at the moment. I just added the screenshots that I felt were necessary. I can remove some if not needed.

Be sure to test a deployment of the app to make sure the SM instance deploys okay. The permanent fix to this archive size limit is to switch to cloning directly from the GH repo rather than seeding CodeCommit with the zip archive. Should be a straight forward change but requires some changes and testing to staging. Regardless, make sure your screenshots/images are sized down to the same/similar res as the others.

File: aws/cloudformation-templates/template.yaml (hunk 26 above)
[the remark this answers is not in the document]
Comment (james-jory) [reply aws/cloudformation-templates/template.yaml:184]: > @james-jory and @ShivaniMehendarge : Do you think it would make sense to also have an option to include/exclude SMS integration for Pinpoint?

I think as long as the app functionality, pre-create lambda, and workshop code has defensive logic when a long code is not available/NONE, we should be fine. For example, we know that with Event Engine all CFN params must have values and that separate long codes can't be pre-provisioned for all accounts ahead of time. So if the event deployer chooses the Pinpoint auto-create, long codes won't be available and SMS functionality should be appropriately disabled. But a missing/NONE long code value should not cause any fatal issues either. However, if later I set a long code in SSM, either manually or in the workshop and redeploy the webapp, the SMS functionality should automatically become active. This is how several other parts of the app work today. For example, the app works fine without Personalize campaigns but once they're created, immediately starts using them. Same with Search and Lex. Make sense?

[the remark this answers is not in the document]
Comment (james-jory) [reply aws/cloudformation-templates/template.yaml:184]: > @alexchirayath I think if we make it optional, it is up to the customers to opt in for SMS integration right? @james-jory I set the default value of the long code to NONE. We are anyways checking that the long code is not NONE in the auto workshop before creating the segment, campaign and template.

This sounds good. As long as the web app, backend, pre-create lambda, and workshops work appropriately with and without a long code, we are in good shape.

[the remark this answers is not in the document]
Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/template.yaml:184]: Yup

**Comment by ShivaniMehendarge** [comment]
> > > Noticed you added some new screenshots to the Pinpoint workshop. Did you test deployment of this branch to make sure the SM notebook instance deploys cleanly? We're already super close to the max 20MB archive size so just want to make sure these new screenshots don't put us over the limit.
> > > Other detailed comments in my review.
> > 
> > 
> > Yes it is 19.1MB at the moment. I just added the screenshots that I felt were necessary. I can remove some if not needed.
> 
> Be sure to test a deployment of the app to make sure the SM instance deploys okay. The permanent fix to this archive size limit is to switch to cloning directly from the GH repo rather than seeding CodeCommit with the zip archive. Should be a straight forward change but requires some changes and testing to staging. Regardless, make sure your screenshots/images are sized down to the same/similar res as the others.

I did test the deployment and verified that the SM instances deploy correctly.

File: aws/cloudformation-templates/template.yaml [hunk 29]
```diff
@@ -174,6 +177,13 @@ Parameters:
       is set to 'Yes'.
     Default: "AWS Retail Demo Store"
   
+  PinpointSMSLongCode:
+    Type: String
+    Description: >
+      Phone number(a dedicated long code) for Amazon Pinpoint to send and receive messages at. Be sure to enable two way SMS for this long code using Pinpoint. This parameter only applies when 'Auto-Configure Pinpoint'
```
Comment (alexchirayath) [inline aws/cloudformation-templates/template.yaml:183]: Can we point them to the workshop/long code aws doc to give them more information?

Comment (ShivaniMehendarge) [reply aws/cloudformation-templates/template.yaml:183]: Sure I can add a link to aws doc

File: src/web-ui/src/partials/TextAlerts/TextAlerts.vue [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+const UsersRepository = RepositoryFactory.get('users');
+
+export default {
+  name: 'TextAlerts',
+    components: {
+      TheMask
+    },
+  data() {
+    return {
+      isEnabled: process.env.VUE_APP_PINPOINT_APP_ID ,
```
Comment (alexchirayath) [inline src/web-ui/src/partials/TextAlerts/TextAlerts.vue:33]: should we called this isPinpointEnabled?

Comment (ShivaniMehendarge) [reply src/web-ui/src/partials/TextAlerts/TextAlerts.vue]: Done

File: src/users/src/users-service/handlers.go [hunk 31]
```diff
@@ -206,18 +206,35 @@ func CreateEndpointAndSendConfirmation(updateEndpointInput *pinpoint.UpdateEndpo
...
+			fmt.Println("Long code not assigned by user");
+			sendMessageInput = &pinpoint.SendMessagesInput {
+				ApplicationId: &pinpoint_app_id,
+				MessageRequest: &pinpoint.MessageRequest {
+					Addresses: sendMessageAddress,
+					MessageConfiguration: &pinpoint.DirectMessageConfiguration {
+						SMSMessage: &pinpoint.SMSMessage {
+							Body: aws.String("Reply Y to receive one time automated marketing messages at this number. No purchase necessary. T&C apply."),
+							MessageType: aws.String("TRANSACTIONAL"),
+						},
```
Comment (james-jory) [inline src/users/src/users-service/handlers.go:220]: So a long code is not required for two-way SMS? The only difference between the "if" and "else" block that I can see is that `OriginationNumber` is included in the else block. Will two-way still work without it and if so why are we dealing with long codes at all?

File: src/web-ui/src/partials/TextAlerts/TextAlerts.vue [hunk 32]
```diff
@@ -17,26 +17,32 @@
...
     };
   },
   computed: {
     ...mapState(['user']),
   },
   methods: {
-    onSubmit() {
+    ...mapActions(['setUser']),
+    async onSubmit() {
       console.log(`Text alerts form submitted with phone number: ${this.phoneNumber}`);
```
Comment (james-jory) [inline src/web-ui/src/partials/TextAlerts/TextAlerts.vue:43]: Please remove

File: src/users/src/users-service/handlers.go (hunk 31 above)
Comment (ShivaniMehendarge) [reply src/users/src/users-service/handlers.go:220]: So the thing is, we are having 2-way SMS. If we dont set up the long code and enable two way SMS for it, customers will receive the confirmation message asking them to  opt in, but when the customers reply with a Y, they won't get any response. Pinpoint does assign a random number to send messages if you dont assign one, it is just for the 2 way SMS that we have to particularly obtain one.  Once you assign the long code, pinpoint won't randomly assign any long/short code, it would pick up the one you have.

File: src/web-ui/src/partials/TextAlerts/TextAlerts.vue (hunk 32 above)
Comment (ShivaniMehendarge) [reply src/web-ui/src/partials/TextAlerts/TextAlerts.vue:43]: Will do.

File: src/users/src/users-service/handlers.go (hunk 31 above)
Comment (james-jory) [reply src/users/src/users-service/handlers.go:220]: So if I understand correctly, the 2-way SMS UX that we're designing here will only work if a long code has been setup, right? If that is the case, should we been sending the opt-in SMS if there is no long code? Won't that create a confusing UX since the customer's reply won't be received and processed as expected?

File: src/web-ui/src/partials/TextAlerts/TextAlerts.vue (hunk 32 above)
Comment (ShivaniMehendarge) [reply src/web-ui/src/partials/TextAlerts/TextAlerts.vue]: Done.

File: src/users/src/users-service/handlers.go (hunk 31 above)
Comment (ShivaniMehendarge) [reply src/users/src/users-service/handlers.go]: Added the long code ssm as well as CFN as discussed offline.

**Comment by alexchirayath** [comment]
Changes look good to me.
Could you please add the disclaimer discussed internally on GitHub as part of the readme (I'd suggest after the  "Getting Started" section)

**Review by alexchirayath — CHANGES_REQUESTED** [review]
Overall LGTM.
Just some minor comments noted below

**Comment by ShivaniMehendarge** [comment]
> Changes look good to me.
> Could you
> 
> * please add the disclaimer discussed internally on GitHub as part of the readme (I'd suggest after the  "Getting Started" section)
> * Resolve the conflicts  on the PR

Added the disclaimer

**Review by alexchirayath — APPROVED** [review]
LGTM, thanks for making the changes. Looks like you have updated the screenshots already so we should be good. If there are any remaining, please add them as well.

**Review by james-jory — CHANGES_REQUESTED** [review]
Just need merge conflicts resolved. Otherwise, looks good.

**Comment by ShivaniMehendarge** [comment]
> Just need merge conflicts resolved. Otherwise, looks good.

@james-jory @alexchirayath It looks like @kazbaig has opened a PR for this with conflicts resolved. (#173 )

**Comment by james-jory** [comment]
Closing since conflicts were rolled up into #173
